shunting-yard.h

```diff
@@ -30,10 +30,13 @@ enum tokType {
   STR, FUNC,
 
   // Numerals:
-  NUM = 0x20,   // Everything with the bit 0x20 set is a number.
-  REAL = 0x21,  // == 0x20 + 0x1 => Real numbers.
-  INT = 0x22,   // == 0x20 + 0x2 => Integral numbers.
-  BOOL = 0x23,  // == 0x20 + 0x3 => Boolean Type.
+  NUM = 0x20,     // Everything with the bit 0x20 set is a number.
+  REAL = 0x21,    // == 0x20 + 0x1 => Real numbers.
+  INT = 0x22,     // == 0x20 + 0x2 => Integral numbers.
+  BOOL = 0x23,    // == 0x20 + 0x3 => Boolean Type.
+  REALREF = 0x24, // == 0x20 + 0x4 => Reference to real numbers.
+  INTREF = 0x25,  // == 0x20 + 0x5 => Reference to integral numbers.
+  BOOLREF = 0x26, // == 0x20 + 0x6 => Reference to boolean Type.
```

Comment:
This is actually one of the biggest concerns: the users of the library only have space to add 17 custom primitive types to the library, with these extra 3 types added to the built-in types we are restricting the number of custom primitive types available to our users to 14. So we really need to discuss that matter of the use-case

Reply:
I understand the concern ... which is going to grow even bigger with the planned addition of reference on string. 

A working solution would be to do add a flag 'EXTREF = 0x100' similar to the existing internal references REF. That way we may handle the references on complex types or any user defined types. I suggest to keep that in mind and defer this change for now,

Right, the discussion around the uses cases is paramount since it invalidates or justifies all these efforts. Adding references on variables to a TokenMap is the way to automagically update these variables whenever they are assigned in CParse. This works also the way around as externally updating the variables affects the outcome of calculations in CParse.  In short, this is a two way bridge between Cparse and its clients.

Reply:
Uhmm, the option of adding `EXTREF = 0x100` would not work because it would use the 9th bit and we are currently using only 8 bits for denoting the type.

Reply:
Hello again @jpfeger,

About the use-case, I am thinking that maybe this might be too much of a specialization to put in the main library.

I am saying that because the library was made to be easy to extend so that each user could specialize to their own needs.

I think this use-case is exactly that type of specialization that will be perfect for more specific use-cases but most users will not use it.

In the other hand, there is a section on the Wiki called "Examples", the changes on this PR are very interesting and I could accept this PR as a separated branch and reference it on the Examples page, this would probably help others and also improve the quaility of the docs.

Do you think this would be a good idea?

Reply:
I was aware of the EXTREF issue. Why limiting ourself to uint_8? I could make the change to uint64_t for oustanding, everlasting coziness. Let me know.

I disagree that the changes are specialized, or with limited general interest. With them, there is a permanent, two way bridge between the clients of the library and the calculations performed by the library. Any external variable passed in reference to a TokenMap is affected by any assignment done to it in CParse at any moment and for any nmber of time. Conversely, any variable`s change gets always reflected in the outcome of any number of CParse calculations at any time.

In my humble opinion, the changes would be quite attractive for any general purpose CParse user and deserve a better fate than sitting in a side branch. You may check with Brandon, he may suggest some adaptations or improvements.

My next goal is to clear the performance issue with the calculations. This is definitively a show stopper for using CParser in our simulation framework. While we are discussing my first contribution, I am going to investigate the issue in parallel.

Reply:
Hello @jpfeger, good news, I spoke to Brandon and he is in favor of adding these changes to the project.

I am sorry for being so hard, and thanks for the suggestion of asking Brandon, I feel better now having a second opinion.

So, in this case, we should proceed to adding the tests and considering where to write the docs (if you want I can write them).

About the tests I think they are testing what needs to be tested but for readability I would suggest changing the ocurrencies of "Pi + a" to the literal value that was expected, i.e.:

```C++
      // ...

      calculator expression("variableB = pi + variableA");
      expression.eval(map);
      EXPECT_TRUE(isEqual(b, 4.14));

      // Demonstrate that the variable 'a' has still an impact on the calculation as it was passed by reference
      a = false;
      expression.eval(map);
      EXPECT_TRUE(isEqual(b, 3.14));
```

What do you think of this change?

Reply:
I am really glad that everything gets ironed out and that we can move forward :-). I am betting that passing references will be the main way to interact with Cparse. 

I agree, the tests are not really edible, so I will propose something better. Same for the documentation. 

I am currently swamped with work and, on top of that, my laptop died on me. Anyway, in the following days, I`ll push the changes for uint64, the references on string, and address your comments.

Reply:
> my laptop died on me.

It's always a pain when that happens :sweat: 

Alright, @jpfeger, If you want I can write the docs, this is a part I find especially important and I like writing docs in general.

Also, you mentioned optimizing the library, maybe you would like to read issue #64 as it relates to one of the main performance issues: We use pointer polymorphism for handling different Token types.

I know that the best JavaScript engines use Unions instead, I was thinking if it would be possible for us to change our tokens to something like this:

```C++
union tokValue_t {
  uint64_t i,
  float64_t f,
  std::string s,
};

// This would be our base token
struct Token {
  tokType_t type
  value tokValue_t
};
```